server/data/data-library/src/main/java/org/apache/james/rrt/lib/CanSendFromImpl.java

```diff
@@ -79,29 +83,48 @@ private boolean emailIsAnAliasOfTheConnectedUser(Username connectedUser, Usernam
             .anyMatch(alias -> alias.equals(connectedUser));
     }
 
-    private Stream<Username> relatedAliases(Username user) throws RecipientRewriteTableException {
-        return Stream.concat(
-            Stream.of(user),
+    private Stream<Username> relatedAliases(Username user) {
+        return recursivelyResolve(user, Throwing.<Username, Stream<Username>>function(targetUser ->
             recipientRewriteTable
-                .listSources(Mapping.alias(user.asString()))
+                .listSources(Mapping.alias(targetUser.asString()))
                 .map(MappingSource::asUsername)
-                .flatMap(OptionalUtils::toStream)
+                .flatMap(OptionalUtils::toStream)).sneakyThrow()
         );
     }
 
     private Stream<Domain> relatedDomains(Username user) {
         return user.getDomainPart()
-            .map(Throwing.function(this::fetchDomains).sneakyThrow())
+            .map(this::fetchDomains)
             .orElseGet(Stream::empty);
     }
 
-    private Stream<Domain> fetchDomains(Domain domain) throws RecipientRewriteTableException {
-        return Stream.concat(
-          Stream.of(domain),
+    private Stream<Domain> fetchDomains(Domain domain) {
+        return recursivelyResolve(domain, Throwing.<Domain, Stream<Domain>>function(targetDomain ->
           recipientRewriteTable
-              .listSources(Mapping.domain(domain))
+              .listSources(Mapping.domain(targetDomain))
               .map(MappingSource::asDomain)
-              .flatMap(OptionalUtils::toStream)
+              .flatMap(OptionalUtils::toStream)).sneakyThrow()
         );
     }
+
+    private <T> Stream<T> recursivelyResolve(T seed, Function<T, Stream<T>> generator) {
```

Comment:
I'm not sure about why you didn't wrote it with a simple recursion, is there a specific reason?

Reply:
Well, it seems to be the same complexity to me.

Reply:
Let me disagree, this small function passes all tests:
```java
    private <T> Stream<T> recursivelyResolve(T seed, Function<T, Stream<T>> generator, int iterations) {
        if (iterations == 0) {
            return Stream.of(seed);
        } else {
            return Stream.concat(
                Stream.of(seed),
                generator.apply(seed).flatMap(item -> recursivelyResolve(item, generator, iterations - 1)));
        }
    }
```

Reply:
Explicit recursion should be avoided, especially in non-TCO languages

Reply:
> Explicit recursions should be avoided, especially in non-TCO languages

Why?

Lack of TCO is a good reason but are there others?

Reply:
Expressiveness, efficiency, showing intent.

Explicit [recursion is the goto of functional programming](https://twitter.com/headinthebox/status/384105824315928577?lang=fr), recursion schemes should be used in order to express intent and to explicit structure.

Reply:
For the record, I found a [good shit in reactor API](https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#expand-java.util.function.Function-) that can replace the hand written StreamSplitter. :trollface: 

But the problem is that there is no support to limit the depth of the expansion.

IMO the current code is really hard to understand so I'm doubling about your `expressiveness`.

Reply:
> For the record, I found a good shit in reactor API that can replace the hand written StreamSplitter. :trollface:

it can be done also

> IMO the current code is really hard to understand so I'm doubling about your `expressiveness`.

See it the other way round, we have moved from `goto`, to `while`/`for`, to end up with `reduce`/`foreach`/etc. each time we have a gain in structure and expressiveness since you specify what you are trying to do.

I did not introduce `anamorphism` and more precise/specialized vocabularies.

Reply:
> IMO the current code is really hard to understand so I'm doubling about your expressiveness.

Could you explain what you think is hard to understand? I think it would help using the right way of writing it for the team

Reply:
I was reading the first version of this PR, basically it looks like a code wrote by a competitive programmer where the problem is resolved but it lacks some steps to abstract the complex logic away. Could be method or class extraction. I saw List, Stream, flatMap, in multiple places, so it overwhelmed me

Reply:
Unnecessary comparison

Reply:
I think the goal of this PR is the contrary: it introduces some functional concept to solve in a classic way (for functional programmers) a given issue.
I admit it is not straightforward for Java developers, and myself had to grin while reading some lines.

As these functional concepts are very interesting and are brought to be used more broadly into our code base, I suggest that @blackheaven takes some time to explain them in general and why he choosed to use them to solve this particular problem.

Reply:
Here is the reason.

When you have a method to write:

```
class Basket {
    private List<Item> items;
    public Float price();
}
```

You have several way to do it.

The imperative style:

```
Float sum = 0.0;
for (Item i : items)
    sum += i.getPrice();

return sum;
```

The functional/declarative:

```
return items
    .streams()
    .mapToFloat(Item::getPrice)
    .sum();
```

Why should we choose on over the other.
After all, the imperative style is more natural, you have no transformation, no type juggling, you know from the first line what you want.

**Answer**: for structure. You have a problem context (which is going from N elements, `Item`, to one, `Float`) which drive your solving strategy.
When you implement a method, you want to clearly state your intent, make your strategy clear.
In order to do that, you use pre-existing strategies leveraging your problem's structure (here, `Collection` to `value`, via `Stream` and `reduce`).
By doing that you ensure to clearly express your intent and having helper methods only focused on well defined actions.

Let's get back to my code.
In James we mostly use `Stream`'s `map` and `reduce`.
For going from a `value` to a `Collection` we do it by hand recursion, which is terrible:

 1. No clear strategy
 2. Elements added in multiple place
 3. Messing/mixing up problem's structure and problem point
 4. Multiplication of big methods

`unfold` is a solution from this kind of problem structure, it deals with it while it's parameters deal with the specificity of the problem.

Reply:
I agree with most of that. However, some points are not objective enough and should not be used as arguments.

> No clear strategy

It doesn't add anything to the demonstration that is otherwise very good. Nothing is objectively "clear".

> Multiplication of big methods

That's not true either. You can very well be recursive and have short methods (my proposed solution was not bad in that respect).

To contribute to what you said, the main points are for me:

1. sharing concepts between people by naming them.
2. build on software community knowledge to be able to rely on related documentation
3. separating the business logic from the technical details

You should learn `fold`, `unfold` and the likes because it makes everything more straightforward to read than hand-crafted algorithm.